### PR TITLE
Handle the deployment with deployers

### PR DESCRIPTION
**Pull request by benoitf**
Fixes https://github.com/eclipse/che/issues/9476
ENV THEIA_PLUGINS can be set for example to
local-dir:/Users/benoitf/theia-plugins

to deploy all plugins from this directory at startup

### CODE REVIEW

**Review by evidolob — APPROVED** [review]

**Review by akurinnoy — APPROVED** [review]

File: packages/plugin-ext/src/main/browser/plugin-frontend-view-contribution.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+            widgetName: 'Plugins',
+            defaultWidgetOptions: {
+                area: 'left',
+                rank: 300
+            },
+            toggleCommandId: 'pluginsView:toggle',
+        });
+    }
+
+    onStart(app: FrontendApplication) {
```
Comment (mmorhun) [inline packages/plugin-ext/src/main/browser/plugin-frontend-view-contribution.ts]: Do we need this empty implementation?

File: packages/plugin-ext/src/main/node/handlers/plugin-vscode-file-handler.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    private unpackedFolder: string;
+    constructor() {
+        this.unpackedFolder = path.resolve(os.tmpdir(), 'vscode-unpacked');
+    }
+
+    accept(resolvedPlugin: PluginDeployerEntry): boolean {
+        return resolvedPlugin.isFile() && resolvedPlugin.path() !== null && resolvedPlugin.path().endsWith(".vsix");
+    }
+
+    async handle(context: PluginDeployerFileHandlerContext): Promise<void> {
```
Comment (mmorhun) [inline packages/plugin-ext/src/main/node/handlers/plugin-vscode-file-handler.ts:25]: Maybe it worsts to move into abstract class, code is identical as in `PluginTheiaFileHandler`? WDYT?

File: packages/plugin-ext/src/main/node/plugin-deployer-directory-handler-context-impl.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { PluginDeployerEntry, PluginDeployerDirectoryHandlerContext } from "../../common/plugin-protocol";
+
+export class PluginDeployerDirectoryHandlerContextImpl implements PluginDeployerDirectoryHandlerContext {
+
+    constructor(private readonly pluginDeployerEntry: PluginDeployerEntry) {
+
+    }
```
Comment (mmorhun) [inline packages/plugin-ext/src/main/node/plugin-deployer-directory-handler-context-impl.ts:14]: Would be good to add empty line.

File: packages/plugin-ext/src/main/browser/plugin-frontend-view-contribution.ts (hunk 1 above)
Comment (benoitf) [reply packages/plugin-ext/src/main/browser/plugin-frontend-view-contribution.ts]: no, will delete

File: packages/plugin-ext/src/main/node/plugin-deployer-directory-handler-context-impl.ts (hunk 3 above)
Comment (benoitf) [reply packages/plugin-ext/src/main/node/plugin-deployer-directory-handler-context-impl.ts:14]: yes

**Comment by mmorhun** [comment]
AS about `local-dir:/Users/benoitf/theia-plugins`. What do you think about using uri's? like `file:///Users/benoitf/theia-plugins` or `local-dir:///Users/benoitf/theia-plugins`. But up to you, just an proposal

File: packages/plugin-ext/src/main/node/handlers/plugin-vscode-file-handler.ts (hunk 2 above)
Comment (benoitf) [reply packages/plugin-ext/src/main/node/handlers/plugin-vscode-file-handler.ts:25]: I will move vscode in another npm module so probably abstract class may be less useful afterwards

**Comment by benoitf** [comment]
yes I think we can change it to be more protocol wise with 3 slashes.
BTW maybe it's working if you use those 3 but I have not tested

File: packages/plugin-ext/src/main/node/handlers/plugin-vscode-file-handler.ts (hunk 2 above)
Comment (mmorhun) [reply packages/plugin-ext/src/main/node/handlers/plugin-vscode-file-handler.ts:25]: ok

**Review by mmorhun — APPROVED** [review]

File: packages/plugin-ext/src/main/browser/plugin-frontend-view-contribution.ts (hunk 1 above)
Comment (benoitf) [reply packages/plugin-ext/src/main/browser/plugin-frontend-view-contribution.ts]: done

File: packages/plugin-ext/src/main/node/plugin-deployer-directory-handler-context-impl.ts (hunk 3 above)
Comment (benoitf) [reply packages/plugin-ext/src/main/node/plugin-deployer-directory-handler-context-impl.ts:14]: done

**Pull request merged**